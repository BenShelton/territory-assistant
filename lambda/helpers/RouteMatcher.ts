import { APIGatewayEvent } from 'aws-lambda'

type APIMethod = 'GET' | 'POST' | 'PUT' | 'DELETE'

export class RouteMatcher {
  private _paramRegex = /:(.*?)(?=(\/|$))/
  private _event: APIGatewayEvent
  private _pathParams: Record<string, string> = {}

  constructor (event: APIGatewayEvent) {
    this._event = event
  }

  get pathParams (): Record<string, string> {
    return this._pathParams
  }

  private _getMatches (path: string, regex: RegExp): string[] {
    const matches = path.match(regex)
    if (!matches) return []
    return matches.slice(1)
  }

  testRoute (method: APIMethod, path: string): boolean {
    if (this._event.httpMethod !== method) return false
    const stringParamRegex = new RegExp(path.replace(this._paramRegex, '(\\w+)') + '/?$')
    if (!stringParamRegex.test(this._event.path)) return false
    const paramNames = this._getMatches(path, this._paramRegex)
    const paramValues = this._getMatches(this._event.path, stringParamRegex)
    this._pathParams = paramNames.reduce((acc: Record<string, string>, p, i) => {
      if (p) acc[p] = paramValues[i]
      return acc
    }, {})
    return true
  }
}
