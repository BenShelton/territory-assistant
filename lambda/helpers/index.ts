import { APIGatewayEvent } from 'aws-lambda'

type APIMethod = 'GET' | 'POST'
type RouteMatcher = (method: APIMethod, path: string) => boolean
interface IResult {
  statusCode: number
  body: string
}

export function generateRouteMatcher (event: APIGatewayEvent): RouteMatcher {
  return function routeMatcher (method: APIMethod, path: string) {
    if (event.httpMethod !== method) return false
    return new RegExp(path + '/?$').test(event.path)
  }
}

export function success<T> (data: T): IResult {
  return { statusCode: 200, body: JSON.stringify(data) }
}

export function badRequest (message: string): IResult {
  return { statusCode: 400, body: message }
}

export function notFound (): IResult {
  return { statusCode: 404, body: 'Unknown Route' }
}
