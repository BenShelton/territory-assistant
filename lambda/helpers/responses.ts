interface IResult {
  statusCode: number
  body: string
}

export function success<T> (data: T): IResult {
  return { statusCode: 200, body: JSON.stringify(data) }
}

export function badRequest (message: string): IResult {
  return { statusCode: 400, body: message }
}

export function unauthorized (): IResult {
  return { statusCode: 401, body: 'Unauthorized' }
}

export function notFound (): IResult {
  return { statusCode: 404, body: 'Unknown Route' }
}
