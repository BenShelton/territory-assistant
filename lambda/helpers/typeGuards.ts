import { IPoint } from 'types'

export function isObject (obj: unknown): obj is Record<string, unknown> {
  return typeof obj === 'object' && !Array.isArray(obj) && !!obj
}

export function isPointList (obj: unknown): obj is IPoint[] {
  if (!Array.isArray(obj)) return false
  const basePoint: IPoint = { lat: 0, lng: 0 }
  return Object.entries(basePoint)
    .every(([k, v]) => {
      return obj.every(p => {
        if (!isObject(p)) return false
        return typeof p[k] === typeof v
      })
    })
}
