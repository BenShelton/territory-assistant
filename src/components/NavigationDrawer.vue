<template>
  <v-navigation-drawer v-model="drawer" app stateless>
    <v-toolbar>
      <v-btn icon @click="closeDrawer">
        <v-icon>mdi-close</v-icon>
      </v-btn>
      <v-toolbar-title>Navigation</v-toolbar-title>
    </v-toolbar>
    <v-list
      v-for="list of navLists"
      :key="list.subheader"
      nav
      subheader
    >
      <v-subheader v-text="list.subheader" />
      <v-list-item v-for="item of list.items" :key="item.title" :to="item.to">
        <v-list-item-icon>
          <v-icon v-text="item.icon" />
        </v-list-item-icon>
        <v-list-item-title v-text="item.title" />
      </v-list-item>
    </v-list>
    <template v-slot:append>
      <v-list nav>
        <v-divider />
        <v-list-item to="/settings">
          <v-list-item-icon>
            <v-icon>mdi-settings</v-icon>
          </v-list-item-icon>
          <v-list-item-title>Settings</v-list-item-title>
        </v-list-item>
        <v-list-item @click="logout">
          <v-list-item-icon>
            <v-icon>mdi-logout-variant</v-icon>
          </v-list-item-icon>
          <v-list-item-title>Logout</v-list-item-title>
        </v-list-item>
      </v-list>
    </template>
  </v-navigation-drawer>
</template>

<script lang="ts">
import Vue from 'vue'

import store from '@/store'

interface INavList {
  subheader: string
  items: {
    to: string
    icon: string
    title: string
  }[]
}

export default Vue.extend({
  name: 'NavigationDrawer',

  computed: {
    drawer: {
      get (): boolean {
        return store.state.drawer.open
      },
      set (val: boolean): void {
        store.commit.drawer.setDrawer(val)
      }
    },
    navLists (): INavList[] {
      return [
        {
          subheader: 'Overview',
          items: [
            { to: '/', icon: 'mdi-home', title: 'Home' },
            { to: '/territory-overview', icon: 'mdi-map', title: 'Territory' }
          ]
        },
        {
          subheader: 'Editors',
          items: [
            { to: '/territory', icon: 'mdi-map', title: 'Territory' },
            { to: '/image', icon: 'mdi-image', title: 'Image' },
            { to: '/maps', icon: 'mdi-map-search', title: 'Maps' },
            { to: '/info', icon: 'mdi-map-marker', title: 'Info' }
          ]
        }
      ]
    }
  },

  watch: {
    '$route.name' () {
      store.commit.drawer.setDrawer(false)
    }
  },

  methods: {
    closeDrawer (): void {
      this.drawer = false
    },
    async logout (): Promise<void> {
      await store.dispatch.auth.logout()
      this.$router.push('/login')
    }
  }
})
</script>
