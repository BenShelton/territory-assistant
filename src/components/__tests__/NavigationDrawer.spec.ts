import { createWrapper } from 'UnitTest/helpers'

import NavigationDrawer from '../NavigationDrawer.vue'

describe('Components/NavigationDrawer', () => {
  it('should have the name "NavigationDrawer"', () => {
    const wrapper = createWrapper(NavigationDrawer, { store: true })
    expect(wrapper.name()).toBe('NavigationDrawer')
  })

  it('should match the snapshot', () => {
    const wrapper = createWrapper(NavigationDrawer, { store: true })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
