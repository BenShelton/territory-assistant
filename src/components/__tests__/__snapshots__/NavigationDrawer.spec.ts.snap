// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/NavigationDrawer should match the snapshot 1`] = `
<nav class="v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile theme--light" style="height: 100vh; top: 0px; transform: translateX(-100%); width: 256px;">
  <div class="v-navigation-drawer__content">
    <header class="v-sheet v-sheet--tile theme--light v-toolbar" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close theme--light"></i></span></button>
        <div class="v-toolbar__title">Navigation</div>
      </div>
    </header>
    <div class="v-list v-sheet v-sheet--tile theme--light v-list--nav"><a href="/" class="v-list-item--active v-list-item v-list-item--link theme--light" tabindex="0">
        <div class="v-list-item__icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-home theme--light"></i></div>
        <div class="v-list-item__title">Home</div>
      </a><a href="/maps" class="v-list-item v-list-item--link theme--light" tabindex="0">
        <div class="v-list-item__icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-map-search theme--light"></i></div>
        <div class="v-list-item__title">Maps</div>
      </a><a href="/settings" class="v-list-item v-list-item--link theme--light" tabindex="0">
        <div class="v-list-item__icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-settings theme--light"></i></div>
        <div class="v-list-item__title">Settings</div>
      </a></div>
  </div>
  <div class="v-navigation-drawer__border"></div>
</nav>
`;
